searxng/searxng
# Simple theme: image detail

## What does this PR do?

One screenshot (the detail are opened on the right, to avoid to hide the thumbnails, they are grouped on the left side of the screen. When the detail is closed, the thumbnails use the whole screen width):
![image](https://user-images.githubusercontent.com/1594191/135746905-f8f168cd-a6f3-47a6-a6f9-e6d9f1755244.png)

TODO (a lot of issues here and there):
* [x] back to top when the detail is hidden when there are only image results.
* [x] ~image selection: sometimes, the blue border doesn't follow the image border (smaller at the bottom on the image)~ related to #368
* [x] detail text: hard coded black & white.
* [x] when the detail is opened, the page offset must be updated to make sure the current selected image is on the screen (the layout reflow may hide the current selected image).
* [x] it is possible to `j` and `k` to go on the next and the previous image, perhaps add some button on screen (?)
* [x] close the detail with the `Esc` key ?
* [x] display a spinner while the JS is loading the full size image.
* [x] define CSS variables from the detail
* [x] the keys `j` and `k` update the detail
* [x] don't hide the search fields at the top of the page (?)
* [x] if the title or description are mangled, display the full text when mouse is over the text
* [x] center the arrows in the next and previous buttons
* [ ] make the touch zone of the next and previous buttons bigger to be usable on a touch screen
* [ ] be able to close the detail modal, if a user presses the currently selected image on the result page (again)

Note: the implementation uses Javascript.
Perhaps it is possible not to use Javascript with anchor ( `#image1` ) and [`:target`](https://developer.mozilla.org/en-US/docs/Web/CSS/:target) CSS selector. It requires to replace [`data-vim-selected`](https://github.com/searxng/searxng/blob/e39a03cc61e9792afb34084fb4d9973a61deecea/searx/static/themes/simple/src/less/style.less#L127-L129) attribute selector with anchors.

## Why is this change important?

Expected behavior / Improve UX.

## How to test this PR locally?

`make themes` to build the static files

Then on a computer:
* `!general !images something`
* `!images something`

On tablet and smartphone, the detail can be full screen (for now I've not tried).

Important:
* the full image size is proxied in the detail. It change the workload on the instance.
* some full image size URL use the `http` protocol: they are blocked by the proxy
* some full image weight more than 5MB: they are blocked too (only when [the `Content-Length` is specified](https://github.com/searxng/searxng/blob/e39a03cc61e9792afb34084fb4d9973a61deecea/searx/webapp.py#L1097-L1101)). The template falls back to the image thumbnail (see https://github.com/searxng/searxng/blob/4599b6cc16dbbaca1949093c512d88aa7e021f2f/searx/static/themes/simple/src/less/style.less#L191-L206 )

## Author's checklist



## Related issues

* map results and embedded players (videos, music) can use the same widget (instead of inline widgets).
